src/System.Security.Principal.Windows/tests/WindowsIdentityTests.cs

```diff
@@ -120,6 +124,57 @@ public static void CheckUserClaims()
         }
     }
 
+    [Fact]
+    public static void RunImpersonatedAsyncTest()
+    {
+        var testData = new RunImpersonatedAsyncTestInfo();
+        BeginTask(testData);
+
+        // Wait for the SafeHandle that was disposed in BeginTask() to actually be closed
+        GC.Collect();
+        GC.WaitForPendingFinalizers();
+        GC.WaitForPendingFinalizers();
+
+        testData.continueTask.Release();
+        testData.task.CheckedWait();
+        if (testData.exception != null)
+        {
+            throw new AggregateException(testData.exception);
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static void BeginTask(RunImpersonatedAsyncTestInfo testInfo)
+    {
+        testInfo.task = null;
+        testInfo.continueTask = new SemaphoreSlim(0, 1);
```

Comment:
Nit: This could just be a TaskCompletionSource.

Reply:
I tried this:

```c#
    [Fact]
    public static void RunImpersonatedAsyncTest()
    {
        var testData = new RunImpersonatedAsyncTestInfo();
        BeginTask(testData);

        // Wait for the SafeHandle that was disposed in BeginTask() to actually be closed
        GC.Collect();
        GC.WaitForPendingFinalizers();
        GC.WaitForPendingFinalizers();

        testData.continueTaskSource.SetResult(true);
        testData.task.CheckedWait();
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void BeginTask(RunImpersonatedAsyncTestInfo testInfo)
    {
        using (SafeAccessTokenHandle token = WindowsIdentity.GetCurrent().AccessToken)
        {
            WindowsIdentity.RunImpersonated(token, () =>
                testInfo.task = Task.Run(async () => await testInfo.continueTaskSource.Task.ConfigureAwait(false)));
        }
    }

    private class RunImpersonatedAsyncTestInfo
    {
        public Task task;
        public TaskCompletionSource<bool> continueTaskSource;

        public RunImpersonatedAsyncTestInfo()
        {
            continueTaskSource = new TaskCompletionSource<bool>();
        }
    }
```

But it doesn't fail in the baseline, don't know why. Adding a Thread.Sleep(1000) before the wait makes it fail. I'm going to leave it as is for now, the current code also allows for unexpected timeouts to be observed and fail the test instead of hanging (though there may be another way to do that).